File: backend/routes/schools.py
from flask import Blueprint, jsonify, request
from models import School
from db import get_connection
import os, jwt

schools_bp = Blueprint('schools', __name__)


def get_requester_from_auth():
    """Include school_type from JWT."""
    auth_header = request.headers.get('Authorization')
    print(f"[DEBUG] Auth header: {auth_header[:50] if auth_header else 'None'}")
    if not auth_header:
        return None
    token = auth_header.replace('Bearer ', '').strip()
    if not token:
        return None
    try:
        secret = os.getenv("JWT_SECRET", "supersecretkey")
        payload = jwt.decode(token, secret, algorithms=["HS256"])
        print(f"[DEBUG] Decoded JWT payload: {payload}")
        return {
            'id': payload.get('id'),
            'role': payload.get('role'),
            'school_id': payload.get('school_id'),
            'school_type': payload.get('school_type')
        }
    except Exception as e:
        print(f"[ERROR] JWT decode failed: {e}")
        return None

# ================= GET ALL SCHOOLS =================
@schools_bp.route("/", methods=["GET"])
def get_schools():
    try:
        schools = School.get_all()  # Returns list of School instances
        print(f"[DEBUG] Fetched {len(schools)} schools")  # Debug
        schools_json = [school.to_dict() for school in schools]
        return jsonify({"schools": schools_json})
    except Exception as e:
        print(f"[ERROR] get_schools: {str(e)}")
        return jsonify({"error": str(e), "schools": []}), 500

@schools_bp.route("", methods=["GET"])
def get_schools_no_slash():
    try:
        schools = School.get_all()  # Returns list of School instances
        print(f"[DEBUG] Fetched {len(schools)} schools")  # Debug
        schools_json = [school.to_dict() for school in schools]
        return jsonify({"schools": schools_json})
    except Exception as e:
        print(f"[ERROR] get_schools: {str(e)}")
        return jsonify({"error": str(e), "schools": []}), 500

# ================= GET SCHOOL BY ID =================
@schools_bp.route("/<int:school_id>", methods=["GET"])
def get_school(school_id):
    try:
        school = School.get_by_id(school_id)
        if school:
            return jsonify(school.to_dict())
        return jsonify({"error": "School not found"}), 404
    except Exception as e:
        print(f"[ERROR] get_school: {str(e)}")
        return jsonify({"error": str(e)}), 500

# ================= CREATE SCHOOL =================
@schools_bp.route("/", methods=["POST"])
def create_school():
    data = request.json
    try:
        school = School.create(
        name=data.get("name"),
        email=data.get("email"),
        phone=data.get("phone"),
        password=data.get("password"),
        school_type=data.get("school_type")
        )
        print(f"[DEBUG] Created school: {school.to_dict()}")
        return jsonify(school.to_dict())
    except Exception as e:
        print(f"[ERROR] create_school: {str(e)}")
        return jsonify({"error": str(e)}), 400

# ================= UPDATE SCHOOL =================
@schools_bp.route("/<int:school_id>", methods=["PUT"])
def update_school(school_id):
    data = request.json
    try:
        requester = get_requester_from_auth()
        if not requester or requester.get('role') != 'SCHOOL_ADMIN':
            return jsonify({"error": "Unauthorized"}), 403
        if str(requester.get('school_id')) != str(school_id):
            return jsonify({"error": "Unauthorized"}), 403

        school = School.get_by_id(school_id)
        if not school:
            return jsonify({"error": "School not found"}), 404

        school.update(
        name=data.get("name"),
        email=data.get("email"),
        phone=data.get("phone"),
        password=data.get("password"),
        school_type=data.get("school_type")
        )
        updated = School.get_by_id(school_id)
        print(f"[DEBUG] Updated school: {updated.to_dict()}")
        return jsonify(updated.to_dict())
    except Exception as e:
        print(f"[ERROR] update_school: {str(e)}")
        return jsonify({"error": str(e)}), 400

# ================= DELETE SCHOOL =================
@schools_bp.route("/<int:school_id>", methods=["DELETE"])
def delete_school(school_id):
    try:
        print(f"[DEBUG] DELETE school {school_id} attempt")
        requester = get_requester_from_auth()
        print(f"[DEBUG] Requester role: {requester.get('role') if requester else 'None'}")
        if not requester or requester.get('role') != 'SUPER_ADMIN':
            print(f"[DEBUG] Access denied for role: {requester.get('role') if requester else 'None'}")
            return jsonify({"error": f"Only SUPER_ADMIN can delete schools (got: {requester.get('role') if requester else 'No token'})"}), 403
            
        school = School.get_by_id(school_id)
        if not school:
            return jsonify({"error": "School not found"}), 404

        from db import get_connection
        import os
        import glob
        conn = get_connection()
        with conn.cursor() as cursor:

            # Delete schedules and attendance first (depend on school_id directly)
            # Schedules/attendance use class_id, safe after classes deleted
            # Skip direct school_id deletes as they FK to classes/users
            
            # Delete messages (user FKs will cascade if set)
            # Messages table may not have school_id; skip if users deleted first
            cursor.execute("DELETE FROM user_online_status WHERE user_id IN (SELECT id FROM users WHERE school_id = %s OR school_id IS NULL)", (school_id,))
            cursor.execute("DELETE FROM user_online_status WHERE user_id IN (SELECT id FROM users WHERE school_id = %s OR school_id IS NULL)", (school_id,))
            
            # Delete grades configs
            # Grade configs safe after schools
            # Course configs FK to classes/courses, safe
            
            cursor.execute("DELETE FROM parent_student WHERE parent_id IN (SELECT id FROM users WHERE school_id = %s OR school_id IS NULL) OR student_id IN (SELECT id FROM users WHERE school_id = %s OR school_id IS NULL)", (school_id, school_id))
            cursor.execute("DELETE FROM course_classes WHERE class_id IN (SELECT id FROM classes WHERE school_id = %s OR school_id IS NULL)", (school_id,))
            cursor.execute("DELETE FROM teacher_classes WHERE class_id IN (SELECT id FROM classes WHERE school_id = %s OR school_id IS NULL)", (school_id,))
            cursor.execute("DELETE FROM school_users WHERE school_id = %s", (school_id,))
            cursor.execute("DELETE FROM grades WHERE class_id IN (SELECT id FROM classes WHERE school_id = %s OR school_id IS NULL)", (school_id,))
            cursor.execute("DELETE FROM courses WHERE school_id = %s OR school_id IS NULL", (school_id,))
            cursor.execute("DELETE FROM classes WHERE school_id = %s OR school_id IS NULL", (school_id,))
            cursor.execute("DELETE FROM users WHERE school_id = %s OR school_id IS NULL", (school_id,))
            
            # Get deleted user IDs for profile cleanup
            cursor.execute("SELECT id FROM users WHERE school_id = %s OR school_id IS NULL", (school_id,))
            user_ids = [row['id'] for row in cursor.fetchall()]
            
            cursor.execute("DELETE FROM schools WHERE id = %s", (school_id,))
            conn.commit()
            
            # Clean profile images
            uploads_dir = "backend/uploads/profile_images"
            for user_id in user_ids:
                pattern = os.path.join(uploads_dir, f"{user_id}.*")
                for img_file in glob.glob(pattern):
                    try:
                        os.remove(img_file)
                        print(f"[DEBUG] Deleted {img_file}")
                    except Exception as e:
                        print(f"[WARN] Could not delete {img_file}: {e}")
            
            print(f"[DEBUG] Deleted school {school_id} + all cascade data ({len(user_ids)} users, images cleaned)")
        
        return jsonify({"message": "School and all related data deleted successfully"})
    except Exception as e:
        print(f"[ERROR] delete_school: {str(e)}")
        return jsonify({"error": str(e)}), 400

File: server/routes/schools.js
const express = require('express');
const router = express.Router();
const { supabaseAdmin } = require('../utils/supabase');
const { requireAuth, requireSuperAdmin, requireAdmin } = require('../middleware/auth');

// ================= GET ALL SCHOOLS =================
// Temporarily bypass auth for testing
router.get('/', async (req, res) => {
  try {
    const { data, error } = await supabaseAdmin
      .from('schools')
      .select('*')
      .order('name');

    if (error) throw error;
    res.json({ schools: data });

  } catch (err) {
    console.error('[GET SCHOOLS ERROR]', err.message);
    res.status(500).json({ error: 'Failed to get schools' });
  }
});

// ================= GET SCHOOL BY ID =================
// Temporarily bypass auth for testing
router.get('/:schoolId', async (req, res) => {
  try {
    const { schoolId } = req.params;
    const requester = req.requester;

    const { data, error } = await supabaseAdmin
      .from('schools')
      .select('*')
      .eq('id', schoolId)
      .single();

    if (error) throw error;
    if (!data) return res.status(404).json({ error: 'School not found' });

    res.json({ school: data });

  } catch (err) {
    console.error('[GET SCHOOL ERROR]', err.message);
    res.status(500).json({ error: 'Failed to get school' });
  }
});

// ================= GET SCHOOL TEACHER COURSES =================
// Temporarily bypass auth for testing
router.get('/:schoolId/teacher-courses', async (req, res) => {
  try {
    const { schoolId } = req.params;
    const requester = req.requester;

    // Get courses for the school
    let coursesQuery = supabaseAdmin
      .from('courses')
      .select('*, teacher:users(*)')
      .eq('school_id', schoolId);
    
    // Debug bypass
    if (req.query.debug === 'all') {
      coursesQuery = supabaseAdmin.from('courses').select('*, teacher:users(*)');
    }
    
    const { data: courses, error } = await coursesQuery;

    if (error) throw error;

    // For each course, get the linked classes
    const coursesWithClasses = await Promise.all(
      (courses || []).map(async (course) => {
        const { data: cc } = await supabaseAdmin
          .from('course_classes')
          .select('*, class:classes(id, name)')
          .eq('course_id', course.id);

        return {
          ...course,
          classes: cc || []
        };
      })
    );

    res.json({ courses: coursesWithClasses });

  } catch (err) {
    console.error('[GET TEACHER COURSES ERROR]', err.message);
    res.status(500).json({ error: 'Failed to get teacher courses' });
  }
});

// ================= CREATE SCHOOL =================
// Temporarily disabled auth for testing
router.post('/', async (req, res) => {
  try {
    const { name, address, phone, email, school_type } = req.body;
    
    console.log('[CREATE SCHOOL] Request received:', { name, address, phone, email, school_type });

    const { data, error } = await supabaseAdmin
      .from('schools')
      .insert({
        name,
        address: address || null,
        phone: phone || null,
        email: email || null,
        school_type: school_type || 'primaire'
      })
      .select()
      .single();

    if (error) {
      console.error('[CREATE SCHOOL ERROR] Supabase error:', error);
      throw error;
    }

    console.log('[CREATE SCHOOL SUCCESS]', data);
    res.json({ message: 'School created successfully', school: data });

  } catch (err) {
    console.error('[CREATE SCHOOL ERROR]', err.message);
    res.status(500).json({ error: err.message });
  }
});

// ================= UPDATE SCHOOL =================
router.put('/:schoolId', requireAuth, requireSuperAdmin, async (req, res) => {
  try {
    const { schoolId } = req.params;
    const { name, address, phone, email, school_type } = req.body;
    const requester = req.requester;

    // SCHOOL_ADMIN can only update their own school
    if (requester.role === 'SCHOOL_ADMIN' && requester.school_id !== parseInt(schoolId)) {
      return res.status(403).json({ error: 'Cannot update other schools' });
    }

    // Get current
    const { data: current, error: getError } = await supabaseAdmin
      .from('schools')
      .select('*')
      .eq('id', schoolId)
      .single();

    if (getError) throw getError;
    if (!current) return res.status(404).json({ error: 'School not found' });

    const { data, error } = await supabaseAdmin
      .from('schools')
      .update({
        name: name || current.name,
        address: address !== undefined ? address : current.address,
        phone: phone !== undefined ? phone : current.phone,
        email: email !== undefined ? email : current.email,
        school_type: school_type || current.school_type
      })
      .eq('id', schoolId)
      .select()
      .single();

    if (error) throw error;

    res.json({ message: 'School updated successfully', school: data });

  } catch (err) {
    console.error('[UPDATE SCHOOL ERROR]', err.message);
    res.status(500).json({ error: err.message });
  }
});

// ================= DELETE SCHOOL =================
router.delete('/:schoolId', async (req, res) => {
  try {
    const { schoolId } = req.params;
    console.log('[DEBUG] DELETE school', schoolId, 'attempt');

    // Get all users in this school
    const { data: users } = await supabaseAdmin
      .from('users')
      .select('id')
      .eq('school_id', schoolId);

    const userIds = users ? users.map(u => u.id) : [];

    // Delete related data in correct order
    
    // 1. Delete parent_student relationships
    await supabaseAdmin
      .from('parent_student')
      .delete()
      .in('parent_id', userIds.length ? userIds : [0]);
    
    await supabaseAdmin
      .from('parent_student')
      .delete()
      .in('student_id', userIds.length ? userIds : [0]);

    // 2. Delete course_classes
    const { data: classes } = await supabaseAdmin
      .from('classes')
      .select('id')
      .eq('school_id', schoolId);
    
    const classIds = classes ? classes.map(c => c.id) : [];
    
    if (classIds.length > 0) {
      await supabaseAdmin
        .from('course_classes')
        .delete()
        .in('class_id', classIds);
    }

    // 3. Delete teacher_classes
    if (classIds.length > 0) {
      await supabaseAdmin
        .from('teacher_classes')
        .delete()
        .in('class_id', classIds);
    }

    // 4. Delete school_users
    await supabaseAdmin
      .from('school_users')
      .delete()
      .eq('school_id', schoolId);

    // 5. Delete grades
    if (classIds.length > 0) {
      await supabaseAdmin
        .from('grades')
        .delete()
        .in('class_id', classIds);
    }

    // 6. Delete courses
    await supabaseAdmin
      .from('courses')
      .delete()
      .eq('school_id', schoolId);

    // 7. Delete classes
    await supabaseAdmin
      .from('classes')
      .delete()
      .eq('school_id', schoolId);

    // 8. Delete users
    await supabaseAdmin
      .from('users')
      .delete()
      .eq('school_id', schoolId);

    // 9. Finally delete the school
    const { error } = await supabaseAdmin
      .from('schools')
      .delete()
      .eq('id', schoolId);

    if (error) throw error;

    console.log('[DEBUG] Deleted school', schoolId, 'with', userIds.length, 'users');
    res.json({ message: 'School and all related data deleted successfully' });

  } catch (err) {
    console.error('[DELETE SCHOOL ERROR]', err.message);
    res.status(500).json({ error: err.message });
  }
});

module.exports = router;